superset/migrations/versions/fe23025b9441_rename_big_viz_total_form_data_fields.py

```
@@ -0,0 +1,92 @@
+from sqlalchemy import Column, Integer, String, Text
+from sqlalchemy.ext.declarative import declarative_base
+
+from superset import db
+
+Base = declarative_base()
+
+
+class Slice(Base):
+    __tablename__ = "slices"
+
+    id = Column(Integer, primary_key=True)
+    params = Column(Text)
+    viz_type = Column(String(250))
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    slices = session.query(Slice).filter(Slice.viz_type == "big_number_total").all()
+    for slc in slices:
+        try:
+            params = json.loads(slc.params)
+            header_format_selector = params.pop("header_format_selector", None)
+            header_timestamp_format = params.pop("header_timestamp_format", None)
+            if header_format_selector:
+                params["force_timestamp_formatting"] = header_format_selector
+            if header_timestamp_format:
+                params["time_format"] = header_timestamp_format
+            slc.params = json.dumps(params, sort_keys=True)
+        except Exception as e:
+            print(e)
+            print(f"Parsing params for slice {slc.id} failed.")
+
+    session.commit()
+    session.close()
+
+
+def downgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+
+    slices = session.query(Slice).filter(Slice.viz_type == "big_number_total").all()
+    for slc in slices:
+        try:
+            params = json.loads(slc.params)
+            time_format = params.pop("time_format", None)
+            force_timestamp_formatting = params.pop("force_timestamp_formatting", None)
+            if time_format:
+                params["header_timestamp_format"] = time_format
+            if force_timestamp_formatting:
+                params["header_format_selector"] = force_timestamp_formatting
+            slc.params = json.dumps(params, sort_keys=True)
+        except Exception as e:
+            print(e)
+            print(f"Parsing params for slice {slc.id} failed.")
```

I've seen a couple migrations that succeeded even when exceptions were raised, and they always had to be fixed later to address inconsistent states.

Here, it would be better to raise the exception, so that the migration fails and the admin has the chance to fix the offending charts before trying again. Otherwise, if the migration succeeds they might not even see the message, leaving their system inconsistent, and even if they see the message they need to downgrade and upgrade, which is more work.

Also, can you update your summary confirming that migrations are atomic, have been tested, and share the benchmark numbers? You can use `scripts/benchmark_migration.py` to get the numbers.

---

Thanks for the comment! I updated the description and and re-raised the exception